pkg/minikube/bootstrapper/kubeadm/kubeadm.go

```diff
@@ -0,0 +1,354 @@
+	// that we also stick in /etc/kubernetes/manifests
+	kubeadmTmpl := "sudo /usr/bin/kubeadm init --config {{.KubeadmConfigFile}} --skip-preflight-checks"
+	t := template.Must(template.New("kubeadmTmpl").Parse(kubeadmTmpl))
+	b := bytes.Buffer{}
+	if err := t.Execute(&b, struct{ KubeadmConfigFile string }{constants.KubeadmConfigFile}); err != nil {
+		return err
+	}
+
+	err := k.c.Run(b.String())
+	if err != nil {
+		return errors.Wrapf(err, "kubeadm init error running command: %s", b.String())
+	}
+
+	//TODO(r2d4): get rid of global here
+	master = k8s.NodeName
+	if err := util.RetryAfter(100, unmarkMaster, time.Millisecond*500); err != nil {
+		return errors.Wrap(err, "timed out waiting to unmark master")
+	}
+
+	if err := util.RetryAfter(100, elevateKubeSystemPrivileges, time.Millisecond*500); err != nil {
+		return errors.Wrap(err, "timed out waiting to elevate kube-system RBAC privileges")
+	}
+
+	return nil
+}
+
+//TODO(r2d4): Split out into shared function between localkube and kubeadm
+func addAddons(files *[]assets.CopyableFile) error {
+	// add addons to file list
+	// custom addons
+	assets.AddMinikubeDirToAssets("addons", constants.AddonsPath, files)
+	// bundled addons
+	for addonName, addonBundle := range assets.Addons {
+		// TODO(r2d4): Kubeadm ignores the kube-dns addon and uses its own.
+		// expose this in a better way
+		if addonName == "kube-dns" {
+			continue
+		}
+		if isEnabled, err := addonBundle.IsEnabled(); err == nil && isEnabled {
+			for _, addon := range addonBundle.Assets {
+				*files = append(*files, addon)
+			}
+		} else if err != nil {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (k *KubeadmBootstrapper) RestartCluster(k8s bootstrapper.KubernetesConfig) error {
+	restoreTmpl := `
+	sudo kubeadm alpha phase certs all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase kubeconfig all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase controlplane all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase etcd local --config {{.KubeadmConfigFile}}
+	`
+	t := template.Must(template.New("restoreTmpl").Parse(restoreTmpl))
+
+	opts := struct {
+		KubeadmConfigFile string
+	}{
+		KubeadmConfigFile: constants.KubeadmConfigFile,
+	}
+
+	b := bytes.Buffer{}
+	if err := t.Execute(&b, opts); err != nil {
+		return err
+	}
+
+	if err := k.c.Run(b.String()); err != nil {
+		return errors.Wrapf(err, "running cmd: %s", b.String())
+	}
+
+	if err := restartKubeProxy(k8s); err != nil {
+		return errors.Wrap(err, "restarting kube-proxy")
+	}
+
+	return nil
+}
+
+func (k *KubeadmBootstrapper) SetupCerts(k8s bootstrapper.KubernetesConfig) error {
+	return bootstrapper.SetupCerts(k.c, k8s)
+}
+
+func (k *KubeadmBootstrapper) UpdateCluster(cfg bootstrapper.KubernetesConfig) error {
+	if cfg.ShouldLoadCachedImages {
+		// Make best effort to load any cached images
+		go machine.LoadImages(k.c, constants.GetKubeadmCachedImages(cfg.KubernetesVersion), constants.ImageCacheDir)
```

Comment:
for my understanding ... do you know that this goroutine will finish before this function returns (do you need to)?

Reply:
No, there no guarantee.  Since this is just a "best efforts" feature. The main function exits when the apiserver is up, so at that point caching and loading is won't have much effect. In practice usually we have more than enough time to cache and load the images